File: typescript/packages/server/src/ssh/stream.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { Writable } from 'node:stream'
import type { ExecuteResult } from '@struktoai/mirage-core/workspace/workspace/types'
import type { ServerChannel } from 'ssh2'

// How far the client may type or pipe ahead of whoever reads it before
// the channel is paused and SSH flow control pushes back.
export const MAX_BUFFERED = 1024 * 1024
export const MAX_LINE = 1024 * 1024
export const MAX_TERMINAL_LINE = 1024

/** A control or input limit delivered in band with channel input. */
export const Mark = {
  EOF: 'eof',
  INTERRUPT: 'interrupt',
  LIMIT: 'limit',
} as const
export type Mark = (typeof Mark)[keyof typeof Mark]

type Item = Uint8Array | Mark

const TAB = 0x09
const LF = 0x0a
const CR = 0x0d
const BS = 0x08
const DEL = 0x7f
const ETX = 0x03
const EOT = 0x04
const NAK = 0x15
const ESC = 0x1b
const CSI_START = 0x5b
const SS3_START = 0x4f
const ERASE = [BS, 0x20, BS]
const NEWLINE = [CR, LF]

export interface DisciplineSink {
  line(bytes: Uint8Array): void
  interrupt(): void
  eof(): void
}

type EscapeState = 'none' | 'start' | 'csi' | 'ss3'

/**
 * The cooked-mode line discipline a pty gives a shell: echo, erase
 * (Backspace, Ctrl-U), Enter, Ctrl-C and Ctrl-D, with escape sequences
 * such as arrow keys swallowed. asyncssh ships one, which the Python door
 * uses; ssh2 does not, so the TypeScript door carries this small one.
 */
export class LineDiscipline {
  private line: number[] = []
  private echoed: number[] = []
  private escape: EscapeState = 'none'
  private afterCR = false

  constructor(
    private readonly echo: (bytes: Uint8Array) => void,
    private readonly sink: DisciplineSink,
  ) {}

  feed(chunk: Uint8Array): void {
    for (const byte of chunk) this.byte(byte)
    if (this.echoed.length > 0) {
      const bytes = Uint8Array.from(this.echoed)
      this.echoed = []
      this.echo(bytes)
    }
  }

  private byte(b: number): void {
    const wasCR = this.afterCR
    this.afterCR = false
    if (this.escape !== 'none') {
      this.skipEscape(b)
      return
    }
    switch (b) {
      case CR:
        this.enter()
        this.afterCR = true
        return
      case LF:
        if (!wasCR) this.enter()
        return
      case BS:
      case DEL:
        this.erase()
        return
      case NAK:
        while (this.line.length > 0) this.erase()
        return
      case ETX:
        this.line = []
        this.sink.interrupt()
        return
      case EOT:
        if (this.line.length === 0) this.sink.eof()
        return
      case ESC:
        this.escape = 'start'
        return
      default:
        break
    }
    if (b < 0x20 && b !== TAB) return
    if (this.line.length >= MAX_TERMINAL_LINE) {
      this.echoed.push(0x07)
      return
    }
    this.line.push(b)
    this.echoed.push(b)
  }

  private enter(): void {
    this.echoed.push(...NEWLINE)
    this.line.push(LF)
    const bytes = Uint8Array.from(this.line)
    this.line = []
    this.sink.line(bytes)
  }

  private erase(): void {
    if (this.line.length === 0) return
    // One code point, not one byte: drop UTF-8 continuation bytes first.
    while (this.line.length > 1 && ((this.line.at(-1) ?? 0) & 0xc0) === 0x80) this.line.pop()
    this.line.pop()
    this.echoed.push(...ERASE)
  }

  private skipEscape(b: number): void {
    if (this.escape === 'start') {
      this.escape = b === CSI_START ? 'csi' : b === SS3_START ? 'ss3' : 'none'
      return
    }
    if (this.escape === 'ss3' || (b >= 0x40 && b <= 0x7e)) this.escape = 'none'
  }
}

/**
 * Everything the client sends on one channel, read once, in order.
 *
 * The channel's data events fill one buffer, so the prompt and the
 * running line's stdin draw from a single ordered stream (typeahead
 * survives a command that did not read it), and an interrupt is seen even
 * while nothing is reading. Ctrl-D ends input for one reader, as a
 * terminal's does; the channel's own EOF ends it for good.
 */
export class ChannelInput {
  private readonly items: Item[] = []
  private buffered = 0
  private closed = false
  private paused = false
  private echoBlocked = false
  private waiters: (() => void)[] = []
  private interruptHandler: (() => void) | null = null
  private readonly discipline: LineDiscipline | null
  private readonly onData = (chunk: Buffer): void => {
    this.received(chunk)
  }
  private readonly onEnd = (): void => {
    this.finish()
  }

  private readonly onDrain = (): void => {
    this.echoBlocked = false
    this.resume()
  }

  constructor(
    private readonly channel: ServerChannel,
    tty: boolean,
  ) {
    this.discipline = tty
      ? new LineDiscipline(
          (bytes) => {
            if (!channel.write(bytes)) {
              this.echoBlocked = true
              this.pause()
            }
          },
          {
            line: (bytes) => {
              this.push(bytes)
            },
            interrupt: () => {
              this.interrupted()
            },
            eof: () => {
              this.push(Mark.EOF)
            },
          },
        )
      : null
  }

  start(): void {
    this.channel.on('data', this.onData)
    this.channel.on('drain', this.onDrain)
    this.channel.on('end', this.onEnd)
    this.channel.on('close', this.onEnd)
  }

  close(): void {
    this.channel.off('data', this.onData)
    this.channel.off('drain', this.onDrain)
    this.channel.off('end', this.onEnd)
    this.channel.off('close', this.onEnd)
    this.finish()
  }

  /**
   * Route Ctrl-C or an INT signal to `handler`. With no handler, the
   * interrupt is queued in band for the prompt to read.
   */
  onInterrupt(handler: (() => void) | null): void {
    this.interruptHandler = handler
  }

  /** A signal request from the client, delivered like Ctrl-C. */
  interruptReceived(): void {
    this.interrupted()
  }

  private received(chunk: Buffer): void {
    // A copy, not the channel's buffer: the bytes may outlive this event
    // in a write buffer, and pyodide refuses a Node Buffer as stdin.
    const bytes = new Uint8Array(chunk)
    if (this.discipline !== null) this.discipline.feed(bytes)
    else this.push(bytes)
  }

  private interrupted(): void {
    if (this.interruptHandler !== null) {
      this.interruptHandler()
      return
    }
    this.push(Mark.INTERRUPT)
  }

  private push(item: Item): void {
    this.items.push(item)
    this.buffered += typeof item === 'string' ? 1 : item.byteLength
    if (this.buffered >= MAX_BUFFERED) this.pause()
    this.wake()
  }

  private took(size: number): void {
    this.buffered -= size
    this.resume()
  }

  private pause(): void {
    if (this.paused) return
    this.paused = true
    this.channel.pause()
  }

  private resume(): void {
    if (this.paused && !this.echoBlocked && this.buffered < MAX_BUFFERED) {
      this.paused = false
      this.channel.resume()
    }
  }

  private finish(): void {
    this.closed = true
    this.wake()
  }

  private wake(): void {
    const waiters = this.waiters
    this.waiters = []
    for (const resolve of waiters) resolve()
  }

  private wait(): Promise<void> {
    return new Promise((resolve) => {
      this.waiters.push(resolve)
    })
  }

  /**
   * The next line, with its newline, or the control that came first: a
   * final unterminated line at the channel's EOF is returned as is,
   * `Mark.INTERRUPT` is a Ctrl-C typed at the prompt, and `Mark.EOF` is
   * Ctrl-D or the channel's EOF, and `Mark.LIMIT` is an oversized line.
   */
  async readline(): Promise<Uint8Array | Mark> {
    const parts: Uint8Array[] = []
    let size = 0
    for (;;) {
      while (this.items.length > 0) {
        const item = this.items[0]
        if (item === undefined) break
        if (typeof item === 'string') {
          if (parts.length > 0) return concat(parts)
          this.items.shift()
          this.took(1)
          return item
        }
        const cut = item.indexOf(LF)
        size += cut < 0 ? item.byteLength : cut
        if (size > MAX_LINE) return Mark.LIMIT
        if (cut < 0) {
          this.items.shift()
          this.took(item.byteLength)
          parts.push(item)
          continue
        }
        this.took(cut + 1)
        parts.push(item.subarray(0, cut + 1))
        const rest = item.subarray(cut + 1)
        if (rest.byteLength > 0) this.items[0] = rest
        else this.items.shift()
        return concat(parts)
      }
      if (this.closed) return parts.length > 0 ? concat(parts) : Mark.EOF
      await this.wait()
    }
  }

  /** The next buffered chunk for a running line's stdin; empty at Ctrl-D or EOF. */
  async read(): Promise<Uint8Array> {
    for (;;) {
      const item = this.items.shift()
      if (typeof item === 'string') this.took(1)
      if (item === Mark.EOF) return new Uint8Array(0)
      if (item !== undefined && typeof item !== 'string') {
        this.took(item.byteLength)
        return item
      }
      if (item === undefined) {
        if (this.closed) return new Uint8Array(0)
        await this.wait()
      }
    }
  }
}

function concat(parts: Uint8Array[]): Uint8Array {
  if (parts.length === 1 && parts[0] !== undefined) return parts[0]
  const out = new Uint8Array(parts.reduce((n, p) => n + p.byteLength, 0))
  let offset = 0
  for (const part of parts) {
    out.set(part, offset)
    offset += part.byteLength
  }
  return out
}

function crlf(data: Uint8Array): Uint8Array {
  let lfs = 0
  for (const b of data) if (b === LF) lfs++
  if (lfs === 0) return data
  const out = new Uint8Array(data.byteLength + lfs)
  let i = 0
  for (const b of data) {
    if (b === LF) out[i++] = CR
    out[i++] = b
  }
  return out
}

function drained(stream: Writable): Promise<void> {
  return new Promise((resolve) => {
    const done = (): void => {
      stream.off('drain', done)
      stream.off('close', done)
      resolve()
    }
    stream.once('drain', done)
    stream.once('close', done)
  })
}

/**
 * Writes a line's output back to the client. On a terminal stderr folds
 * into stdout, as a pty points both descriptors at one device, and each
 * newline goes out as CRLF, as asyncssh's line editor does on the Python
 * side.
 */
export class ChannelOutput {
  constructor(
    private readonly channel: ServerChannel,
    private readonly tty: boolean,
  ) {}

  async write(data: Uint8Array, stderr = false): Promise<void> {
    if (data.byteLength === 0) return
    const stream: Writable = stderr && !this.tty ? this.channel.stderr : this.channel
    if (!stream.writable) return
    if (!stream.write(this.tty ? crlf(data) : data)) await drained(stream)
  }
}

/** The channel's input as a line's stdin: chunks until Ctrl-D or EOF. */
export async function* channelStdin(source: ChannelInput): AsyncGenerator<Uint8Array> {
  for (;;) {
    const chunk = await source.read()
    if (chunk.byteLength === 0) return
    yield chunk
  }
}

/** A line's stdout, then its stderr, onto the channel. */
export async function deliver(result: ExecuteResult, output: ChannelOutput): Promise<void> {
  await output.write(result.stdout)
  await output.write(result.stderr, true)
}

File: python/mirage/server/ssh/stream.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
import logging
from collections import deque
from collections.abc import Awaitable, Callable
from enum import Enum

import asyncssh
from asyncssh.editor import SSHLineEditorChannel

from mirage.io.cooperative import chunks
from mirage.io.types import ByteSource

logger = logging.getLogger(__name__)

ENCODING = "utf-8"
# asyncssh runs its line editor only on a text channel, and
# surrogateescape is the one error policy under which every byte
# survives the str round trip, so binary input and output cross a text
# channel unchanged.
ERRORS = "surrogateescape"
READ_SIZE = 64 * 1024
# How far the client may type or pipe ahead of whoever reads it before
# the channel stops being read and SSH flow control pushes back.
MAX_BUFFERED = 1024 * 1024
MAX_LINE = 1024 * 1024
MAX_TERMINAL_LINE = 1024

Send = Callable[[bytes, bool], Awaitable[None]]


class Mark(Enum):
    """A control or input limit delivered in band with channel input."""

    EOF = "eof"
    INTERRUPT = "interrupt"
    LIMIT = "limit"


def encode(text: str) -> bytes:
    return text.encode(ENCODING, ERRORS)


def decode(data: bytes) -> str:
    return data.decode(ENCODING, ERRORS)


class ChannelInput:
    """Everything the client sends on one channel, read once, in order.

    One pump task reads the channel into a buffer, so the prompt and the
    running line's stdin draw from a single ordered stream (typeahead
    survives a command that did not read it), and an interrupt is seen
    even while nothing is reading. Ctrl-D from asyncssh's line editor
    ends input for one reader, as a terminal's does; the channel's own
    EOF ends it for good.

    Args:
        process (asyncssh.SSHServerProcess[str]): the channel's process.
    """

    def __init__(self, process: asyncssh.SSHServerProcess[str]) -> None:
        self._process = process
        self._items: deque[bytes | Mark] = deque()
        self._buffered = 0
        self._closed = False
        self._changed = asyncio.Event()
        self._room = asyncio.Event()
        self._room.set()
        self._interrupt: Callable[[], None] | None = None
        self._task: asyncio.Task[None] | None = None

    def start(self) -> None:
        self._task = asyncio.create_task(self._pump())

    async def close(self) -> None:
        if self._task is None:
            return
        self._task.cancel()
        await asyncio.wait([self._task])
        if not self._task.cancelled() and self._task.exception():
            logger.warning("ssh: input pump failed: %r",
                           self._task.exception())

    def on_interrupt(self, handler: Callable[[], None] | None) -> None:
        """Route Ctrl-C (a break) or an INT signal to ``handler``.

        With no handler, the interrupt is queued in band for the prompt
        to read, and the line editor's half-typed input is dropped.

        Args:
            handler (Callable[[], None] | None): called on the channel's
                loop for each interrupt, or None to queue them.
        """
        self._interrupt = handler

    async def _pump(self) -> None:
        stdin = self._process.stdin
        while True:
            await self._room.wait()
            try:
                data = await stdin.read(READ_SIZE)
            except (asyncssh.BreakReceived, asyncssh.SignalReceived):
                self._interrupted()
                continue
            except asyncssh.TerminalSizeChanged:
                continue
            except (asyncssh.Error, OSError) as exc:
                logger.debug("ssh: channel input ended: %r", exc)
                break
            if data:
                self._push(encode(data))
            elif stdin.at_eof():
                break
            else:
                self._push(Mark.EOF)
        self._closed = True
        self._changed.set()

    def _interrupted(self) -> None:
        if self._interrupt is not None:
            self._interrupt()
            return
        chan = self._process.channel
        if isinstance(chan, SSHLineEditorChannel):
            chan.clear_input()
        self._push(Mark.INTERRUPT)

    def _push(self, item: bytes | Mark) -> None:
        self._items.append(item)
        self._buffered += len(item) if isinstance(item, bytes) else 1
        if self._buffered >= MAX_BUFFERED:
            self._room.clear()
        self._changed.set()

    def _took(self, size: int) -> None:
        self._buffered -= size
        if self._buffered < MAX_BUFFERED:
            self._room.set()

    async def _wait(self) -> None:
        self._changed.clear()
        await self._changed.wait()

    async def readline(self) -> bytes | Mark:
        """The next line, with its newline, or the control that came first.

        Returns:
            bytes | Mark: the line (a final unterminated one at the
                channel's EOF is returned as is), ``Mark.INTERRUPT`` for
                a Ctrl-C typed at the prompt, or ``Mark.EOF`` for Ctrl-D
                or the channel's EOF; ``Mark.LIMIT`` for an oversized line.
        """
        line = bytearray()
        while True:
            while self._items:
                item = self._items[0]
                if isinstance(item, Mark):
                    if line:
                        return bytes(line)
                    self._items.popleft()
                    self._took(1)
                    return item
                cut = item.find(b"\n")
                size = len(item) if cut < 0 else cut
                if len(line) + size > MAX_LINE:
                    return Mark.LIMIT
                if cut < 0:
                    self._items.popleft()
                    self._took(len(item))
                    line += item
                    continue
                self._took(cut + 1)
                line += item[:cut + 1]
                rest = item[cut + 1:]
                if rest:
                    self._items[0] = rest
                else:
                    self._items.popleft()
                return bytes(line)
            if self._closed:
                return bytes(line) if line else Mark.EOF
            await self._wait()

    async def read(self) -> bytes:
        """The next buffered chunk, for a running line's stdin.

        Returns:
            bytes: the chunk, or ``b""`` at Ctrl-D or the channel's EOF.
        """
        while True:
            if self._items:
                item = self._items.popleft()
                if isinstance(item, Mark):
                    self._took(1)
                if item is Mark.EOF:
                    return b""
                if isinstance(item, bytes):
                    self._took(len(item))
                    return item
                continue
            if self._closed:
                return b""
            await self._wait()


class ChannelOutput:
    """Writes a line's output back to the client.

    On a terminal stderr folds into stdout, as a pty points both
    descriptors at one device; asyncssh's line editor then turns each
    newline into CRLF and redraws the input line around the output.

    Args:
        process (asyncssh.SSHServerProcess[str]): the channel's process.
        tty (bool): whether the client asked for a terminal.
    """

    def __init__(self, process: asyncssh.SSHServerProcess[str],
                 tty: bool) -> None:
        self._process = process
        self._tty = tty

    async def write(self, data: bytes, stderr: bool = False) -> None:
        stream = (self._process.stderr
                  if stderr and not self._tty else self._process.stdout)
        stream.write(decode(data))
        await stream.drain()


class LoopStdin:
    """A channel's input as a line's stdin, pulled from the workspace loop.

    The line runs on the workspace runner's loop while the channel lives
    on the daemon's, so each pull hops to the daemon loop for the next
    buffered chunk and waits there.

    Args:
        source (ChannelInput): the channel's input.
        loop (asyncio.AbstractEventLoop): the loop the channel lives on.
    """

    def __init__(self, source: ChannelInput,
                 loop: asyncio.AbstractEventLoop) -> None:
        self._source = source
        self._loop = loop

    def __aiter__(self) -> "LoopStdin":
        return self

    async def __anext__(self) -> bytes:
        data = await asyncio.wrap_future(
            asyncio.run_coroutine_threadsafe(self._source.read(), self._loop))
        if not data:
            raise StopAsyncIteration
        return data


def loop_sender(output: ChannelOutput,
                loop: asyncio.AbstractEventLoop) -> Send:
    """A writer the workspace loop can await, landing on the channel's loop.

    Awaiting it waits for the channel to drain, so a fast producer is
    paced by the client instead of buffered without bound.

    Args:
        output (ChannelOutput): the channel's output.
        loop (asyncio.AbstractEventLoop): the loop the channel lives on.

    Returns:
        Send: ``send(data, stderr)``.
    """

    async def send(data: bytes, stderr: bool) -> None:
        await asyncio.wrap_future(
            asyncio.run_coroutine_threadsafe(output.write(data, stderr), loop))

    return send


async def deliver(stdout: ByteSource | None, stderr: ByteSource | None,
                  send: Send) -> None:
    """Stream a line's stdout, then its stderr, through ``send``.

    Args:
        stdout (ByteSource | None): the line's stdout.
        stderr (ByteSource | None): the line's stderr.
        send (Send): where each chunk goes.
    """
    for source, is_stderr in ((stdout, False), (stderr, True)):
        if source is None:
            continue
        async for chunk in chunks(source):
            if chunk:
                await send(chunk, is_stderr)
